examples/jsm/nodes/core/InputNode.js

```diff
@@ -37,51 +37,69 @@ function getValueType( value ) {
 
 	}
 
+	return null;
+
 }
 
 function getValueFromType( type ) {
 
+	const last4 = type?.slice( -4 );
+
 	if ( type === 'color' ) {
 
 		return new Color();
 
-	} else if ( type.includes( 'vec2' ) ) {
+	} else if ( last4 === 'vec2' ) {
 
 		return new Vector2();
 
-	} else if ( type.includes( 'vec3' ) ) {
+	} else if ( last4 === 'vec3' ) {
 
 		return new Vector3();
 
-	} else if ( type.includes( 'vec4' ) ) {
+	} else if ( last4 === 'vec4' ) {
 
 		return new Vector4();
 
-	} else if ( type.includes( 'mat3' ) ) {
+	} else if ( last4 === 'mat3' ) {
 
 		return new Matrix3();
 
-	} else if ( type.includes( 'mat4' ) ) {
+	} else if ( last4 === 'mat4' ) {
 
 		return new Matrix4();
 
 	}
 
+	return null;
+
 }
 
 class InputNode extends Node {
 
-	constructor( value, nodeType ) {
+	constructor( value, nodeType = null ) {
 
-		if ( nodeType === undefined ) {
+		super( nodeType );
 
-			nodeType = getValueType( value );
+		this.value = value;
 
-		}
+	}
 
-		super( nodeType );
+	get value() {
 
-		this.value = value;
+		return this._value;
+
+	}
+
+	set value( val ) {
+
+		this._value = val;
+
+		if ( this.nodeType === null ) {
+
+			this.nodeType = getValueType( val );
```

Comment:
I think it would be better in `.getNodeType()`. This check is to be done only in `NodeBuilder` for unnecessary processing and preserve the serializing. This approach opens up other problems too like this as well:

```js
const myUniform = new UniformNode();
myUniform.value = 1;
myUniform.value = new Vector3(); // nodeType will no longer be null
```

Reply:
this.nodeType is also used in serialization/deserialization. I do not see why `uniform.value = 1; uniform.value = new Vector3()` pattern could be ever used.

Reply:
> this.nodeType is also used in serialization/deserialization. 

`nodeType` should preserve the value defined by the user here `null == auto` in serialization, we have `getNodeType()` to define the dynamic type. 


> I do not see why uniform.value = 1; uniform.value = new Vector3() pattern could be ever used.

`UniformNode` is a transporter, it must adapt. It's not time to get the real `nodeType` in `set/get`, we're confusing stages here.

see `Object3DNode` it's set a `.value` per frame, `set/get` is not the most optimized way to deal with it, I think we need to fix this before merger.

Reply:
OK, is it better now?

Reply:
Yes, yess! thanks!

Reply:
You can write `data.nodeType = this.getNodeType()` but it could possibly break if a child of InputNode redefined `.getNodeType()`. That is why I created a new function `._getNodeType()`.